power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/dataset.hpp

```diff
@@ -4,117 +4,263 @@
+static_assert(is_data_mutable_v<mutable_dataset_t>);
+static_assert(is_data_mutable_v<writable_dataset_t>);
+static_assert(!is_indptr_mutable_v<const_dataset_t>);
+static_assert(!is_indptr_mutable_v<mutable_dataset_t>);
+static_assert(is_indptr_mutable_v<writable_dataset_t>);
+
+struct ComponentInfo {
+    MetaComponent const* component;
+    // for non-uniform component, this is -1, we use indptr to describe the elements per scenario
+    Idx elements_per_scenario;
+    Idx total_elements;
+};
+
+struct DatasetInfo {
+    bool is_batch;
+    Idx batch_size; // for single dataset, the batch size is one
+    MetaDataset const* dataset;
+    std::vector<ComponentInfo> component_info;
+};
+
+template <dataset_type_tag dataset_type_> class Dataset {
+    struct immutable_t {};
+    struct mutable_t {};
 
-template <dataset_type_tag dataset_type_> class DataPointer {
   public:
     using dataset_type = dataset_type_;
+    using Data = std::conditional_t<is_data_mutable_v<dataset_type>, void, void const>;
+    using Indptr = std::conditional_t<is_indptr_mutable_v<dataset_type>, Idx, Idx const>;
 
-  private:
-    template <class T> using ptr_t = std::conditional_t<is_const_dataset_v<dataset_type>, T const*, T*>;
+    template <class StructType>
+    using DataStruct = std::conditional_t<is_data_mutable_v<dataset_type>, StructType, StructType const>;
 
-  public:
-    DataPointer() : ptr_{nullptr}, indptr_{nullptr}, batch_size_{}, elements_per_scenario_{} {}
-
-    // single batch dataset
-    DataPointer(ptr_t<void> ptr, Idx single_length)
-        : ptr_{ptr}, indptr_{nullptr}, batch_size_{1}, elements_per_scenario_{single_length} {}
-
-    // fix batch length
-    DataPointer(ptr_t<void> ptr, Idx batch_size, Idx elements_per_scenario)
-        : ptr_{ptr}, indptr_{nullptr}, batch_size_{batch_size}, elements_per_scenario_{elements_per_scenario} {}
-
-    // variable batches
-    DataPointer(ptr_t<void> ptr, Idx const* indptr, Idx batch_size)
-        : ptr_{ptr}, indptr_{indptr}, batch_size_{batch_size}, elements_per_scenario_{-1} {}
-
-    // copy to const constructor
-    DataPointer(ptr_t<void> ptr, Idx const* indptr, Idx batch_size, Idx elements_per_scenario)
-        : ptr_{ptr}, indptr_{indptr}, batch_size_{batch_size}, elements_per_scenario_{elements_per_scenario} {}
-
-    template <class T> std::pair<ptr_t<T>, ptr_t<T>> get_iterators(Idx pos) const {
-        assert(pos < batch_size_);
-        auto* const ptr = reinterpret_cast<ptr_t<T>>(ptr_);
-        if (indptr_) {
-            if (pos < 0) {
-                return std::make_pair(ptr, ptr + indptr_[batch_size_]);
-            }
-            return std::make_pair(ptr + indptr_[pos], ptr + indptr_[pos + 1]);
+    struct Buffer {
+        Data* data;
+        // for uniform buffer, indptr is empty
+        std::span<Indptr> indptr;
+    };
+
+    static constexpr Idx invalid_index{-1};
+
+    Dataset(bool is_batch, Idx batch_size, std::string_view dataset, MetaData const& meta_data)
```

Comment:
providing `.is_batch = false, .batch_size = 1` feels redundant to me. Shouldn't we provide overloads like so?
```suggestion
    Dataset(Idx batch_size, std::string_view dataset, MetaData const& meta_data);
    Dataset(std::string_view dataset, MetaData const& meta_data);
```
?